core/src/Streamly/Internal/Control/Exception.hs

```diff
@@ -165,54 +186,110 @@ newtype AllocateIO = AllocateIO
-                ((i + 1, Map.insert i (void $ free r) mp), i))
-            return (r, idx)
-
-        let modify (i, mp) =
-                let res = Map.lookup index mp
-                 in ((i, Map.delete index mp), res)
-
-            release = do
-                f <- atomicModifyIORef ref modify
-                sequence_ f
+    -- XXX can we ensure via GC that the resources that we are freeing are all
+    -- dead, there are no other references to them?
+    --
+    -- XXX Take a lock to avoid parallel release. By using an MVar for each
+    -- "release" instance we can make release thread-safe such that it can be
+    -- called from multiple threads concurrently. We can use acquireSafeRelease
+    -- to optionally use that feature. Using that will make release
+    -- interruptible. Also using a shared resource in multiple threads may have
+    -- more issues like one thread releasing it means other threads should not
+    -- use it any more.
+    --
+    -- XXX We can use atomicModifyIORef to run each release action only once,
+    -- in one of the threads instead of using an MVar.
+
+    aft ref =
+        -- XXX free them atomically, even if another release executes in
+        -- parallel.
+        liftIO $ mask_ $ do
+            (_, mp1, _) <- readIORef ref
+            -- liftIO $ putStrLn "cleaning up priority 1"
+            -- Note that the channel cleanup function is interruptible because
+            -- it has blocking points.
+            sequence_ mp1
+            -- Now nobody would be changing mp2, we can read it safely
+            (_, _, mp2) <- readIORef ref
+            -- liftIO $ putStrLn "cleaning up priority 2"
+            sequence_ mp2
+
+    allocator ref pri alloc free = do
+        let insertResource r (i, mp1, mp2) =
+                case pri of
+                    Priority1 ->
+                        ((i + 1, Map.insert i (void $ free r) mp1, mp2), i)
+                    Priority2 ->
+                        ((i + 1, mp1, Map.insert i (void $ free r) mp2), i)
+
+        (r, index) <-
+            liftIO $ mask_ $ do
+                -- tid <- myThreadId
+                r <- alloc
+                idx <- atomicModifyIORef' ref (insertResource r)
+                -- liftIO $ putStrLn $ "insert: " ++ show pri
+                --      ++ " " ++ show idx ++ " " ++ show tid
+                return (r, idx)
+
+        let deleteResource (i, mp1, mp2) =
+                case pri of
+                    Priority1 ->
+                        let res = Map.lookup index mp1
+                         in ((i, Map.delete index mp1, mp2), res)
+                    Priority2 ->
+                        let res = Map.lookup index mp2
+                         in ((i, mp1, Map.delete index mp2), res)
+
+            release =
+                -- IMPORTANT: do not use interruptible operations in this
+                -- critical section. Even putStrLn can make tests fail.
```

Comment:
What exactly are interruptible operations?
Operations that allow the thread to sleep?